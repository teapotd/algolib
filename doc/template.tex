\documentclass[10pt]{article}

\usepackage[T1]{fontenc}
\usepackage{geometry}
\usepackage{multicol}
\usepackage{titletoc}
\usepackage{courier}
\usepackage{fancyhdr}
\usepackage{minted}
\usepackage{ulem}
\usepackage{graphicx}
\usepackage[hidelinks]{hyperref}

\geometry{
	a4paper,
	landscape,
	includehead,
	hmargin={0.5cm,0.5cm},
	vmargin={0.5cm,0.6cm},
	headsep=0.1cm,
	footskip=0.25cm
}

\setlength{\columnsep}{8pt}
\setlength{\parindent}{0pt}
\setlength{\columnseprule}{0.5pt}
\renewcommand{\footrulewidth}{0.5pt}

\pagestyle{fancy}
\fancyhf{}
\lhead{Jagiellonian University - Jagiellonian 1}
\rhead{\thepage}

\makeatletter

% Minted style settings
\setminted{tabsize=2}
\setminted{fontsize=\scriptsize}
\setminted{mathescape=true}
\usemintedstyle{tango}

% Fixes space length in minted environment, but disables underlines globally. TODO: do something smarter
\let\underline\relax

% Minimalistic style for table of contents
\renewcommand\tableofcontents{\@starttoc{toc}}
\titlecontents{section}[0em]{}{\thecontentslabel}{\thecontentslabel}{\hfill\contentspage}

% Make empty lines in code listings smaller
\let\FV@ListProcessLineOriginal\FV@ListProcessLine
\def\FV@ListProcessLine#1{%
  \ifx\FV@Line\empty%
    \hbox{}\vspace{-4pt}%
  \else%
    \FV@ListProcessLineOriginal{#1}%
  \fi}

% Environment for code listing. Arguments: title, language, hash.
\newenvironment{code}[3]{%
	\uline{\textbf{\Verb{#1}}\hfill\Verb{#3}}%
	\phantomsection%
	\addcontentsline{toc}{section}{\protect\Verb{#1}}%
	\VerbatimEnvironment%
	\let\FV@ListVSpace\relax%
	\vspace*{4pt}%
	\begin{minted}{#2}}
	{\end{minted}}

\makeatother

\begin{document}

\begin{multicols*}{4}
	\tableofcontents
	\vspace{\fill}\pagebreak
	{CONTENT}
\end{multicols*}

\begin{center}
	\includegraphics[scale=1.185]{../doc/appendix.png}
\end{center}

\end{document}
