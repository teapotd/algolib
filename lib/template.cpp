#pragma GCC optimize("Ofast,no-stack-protector")

#include <bits/stdc++.h>
using namespace std;
using namespace rel_ops;

using ll  = int64_t;
using Vi  = vector<int>;
using Pii = pair<int, int>;

#define mp make_pair

#define rep(i, b, e)  for (int i = (b); i <  (e); i++)
#define repd(i, b, e) for (int i = (b); i >= (e); i--)
#define each(a, x)    for (auto& a : (x))
#define all(x)        (x).begin(), (x).end()
#define sz(x)         int((x).size())

// Debugging

#define TM template<class T, class U=int, class ...V>
void dbgx(...) {}

TM void dbgx(T s, U a, V... b) {
	while (*s && *s != ',') cerr << *s++;
	cerr << ": " << a << *s++; dbgx(s, b...);
}

#ifdef LOC
#define dbg(...) (cerr << "<" << __LINE__, dbgx("> " #__VA_ARGS__, __VA_ARGS__), cerr << "\n")
#else
#define dbg(...)
#endif

#define DF(t, ...) TM auto operator<<(ostream& o, __VA_ARGS__ x)->decltype(t,cout)

DF(&T::print, T)  { x.print(); return o; }
DF(0, pair<T, U>) { return o << "(" << x.first << ", " << x.second << ")"; }

DF(T().begin(), T) {
	o << "[";
	each(e, x) o << e << ", ";
	return o << "]";
}

#define DD(...) void print() { cerr << '{'; dbgx(#__VA_ARGS__, __VA_ARGS__); cerr << '}'; }

// ---

int main() {
	cin.sync_with_stdio(0); cin.tie(0);
	cout << fixed << setprecision(10);
	return 0;
}
