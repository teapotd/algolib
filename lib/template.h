#pragma once

#include <bits/stdc++.h>
using namespace std;

using ll    = int64_t;
using ull   = uint64_t;
using ld    = long double;
using cmpl  = complex<double>;

#define IT iterator

#define rep(i, b, e)  for (int i = b; i < (e); i++)
#define repd(i, b, e) for (int i = int(b); i >= (e); i--)
#define each(a, x)    for (auto& a : x)
#define all(x)        (x).begin(), (x).end()
#define sz(x)         int((x).size())


// unique_ptr without deallocation
template<typename T>
struct single_ptr {
	T* elem{0};

	single_ptr()                              {}
	single_ptr(nullptr_t)                     {}
	single_ptr(T* v)           : elem(v)      {}
	single_ptr(single_ptr&& r) : elem(r.elem) { r.elem = 0; }

	single_ptr& operator=(nullptr_t)      { elem = 0; return *this; }
	single_ptr& operator=(single_ptr&& r) { elem = r.elem; r.elem = 0; return *this; }

	T* operator->() { return elem; }
	T& operator*()  { return *elem; }
	operator bool() { return elem; }
};


// #!/bin/bash
// set -e


// Normal build:
// g++ -O2 -Wall -Wextra -std=c++11 -o $1.e $1.cpp


// Debug build:
// g++-6 -fsanitize=address -fsanitize=undefined -D_GLIBCXX_DEBUG -O2 -Wall -Wextra -std=c++11 -o $1.e $1.cpp
// -fvisibility=hidden (chyba tylko mac)


// Stack limit:
// -Wl,-stack_size -Wl,16000000 -Wl,-no_pie


// Compare:
// ./build.sh $1; ./build.sh $2; ./build.sh $3
//
// while :; do
// 	 $3.e > cmp.in; echo -n 0
// 	 $1.e < cmp.in > prog1.out; echo -n 1
// 	 $2.e < cmp.in > prog2.out; echo -n 2
// 	 diff prog1.out prog2.out
// 	 echo -n Y
// done
