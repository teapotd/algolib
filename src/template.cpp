#include <bits/stdc++.h>
using namespace std;

using ll  = long long;
using Vi  = vector<int>;
using Pii = pair<int,int>;

#define mp make_pair
#define pb push_back
#define x  first
#define y  second

#define rep(i,b,e) for(int i=(b); i<(e); i++)
#define each(a,x)  for(auto& a : (x))
#define all(x)     (x).begin(),(x).end()
#define sz(x)      int((x).size())

int main() {
	cin.sync_with_stdio(0); cin.tie(0);
	cout << fixed << setprecision(12);

	// Don't call destructors:
	cout << flush; _Exit(0);
}

// > Debug printer

#define tem \
	template<class t,class u,class...w> auto
#define pri(x,y,z) \
	tem operator<<(t&o,u a)->decltype(z,o) \
	{ o << *#x; y; z; return o << #x+1; }

pri({},, a.print())
pri((),, o << a.x << ", " << a.y)

pri([], auto d=""; for (auto i : a)
	(o << d << i, d = ", "), all(a))

void DD(...) {}
tem DD(t s, u a, w... k) {
	for (int b=1; cerr << *s++, *s && *s - b*44;)
		b += 2 / (*s*2 - 81);
	cerr << ": " << a; DD(s, k...);
}

#ifdef LOC
#define deb(x...) \
	DD("[,\b :] "#x, __LINE__, x), cerr << endl
#else
#define deb(...)
#endif

#define DBP(x...) void print() { DD(#x, x); }

// > Utils

// Return smallest k such that 2^k > n
// Undefined for n = 0!
int uplg(int n) { return 32-__builtin_clz(n); }
int uplg(ll n){ return 64-__builtin_clzll(n); }

// Compare with certain epsilon (branchless)
// Returns -1 if a < b; 1 if a > b; 0 if equal
// a and b are assumed equal if |a-b| <= eps
int cmp(double a, double b, double eps=1e-9) {
	return (a > b+eps) - (a+eps < b);
}
