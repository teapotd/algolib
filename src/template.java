import java.io.*;
import java.util.*;

public class Task extends PrintWriter {
	BufferedReader reader = new BufferedReader(
			new InputStreamReader(System.in), 32768);
	StringTokenizer tok;

	public static void main(String[] a) {
		try (Task t = new Task()) { t.solve(); }
	}

	Task() { super(System.out); }

	String scan() {
		while (tok == null || !tok.hasMoreTokens())
			try {
				tok = new StringTokenizer(
					reader.readLine());
			} catch (Exception e) {
				throw new RuntimeException(e);
			}
		return tok.nextToken();
	}

	int scanInt() {
		return Integer.parseInt(scan());
	}

	void solve() {
		int n = scanInt();
		printf("hello %d", n);
	}
}
