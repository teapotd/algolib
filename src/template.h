#pragma once
//!!EXCLUDE-FILE

#ifdef __clang__
#define private public
#define _M_p __seg_
#endif

#define main __main_defined_by_template
#include "template.cpp"
#undef main
