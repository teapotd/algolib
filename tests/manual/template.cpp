#include "testing.h"
#include "../../src/template.h"

struct Vert {
	int weight;
	vi edges;

	DBP(weight, edges);
};

int main() {
	vector<Vert> graph;

	rep(i, 0, 10) {
		Vert v;
		v.weight = i + 100;
		rep(j, 0, i+2) v.edges.pb(i+j);
		graph.pb(v);
	}

	deb(graph);
	rep(i, 0, sz(graph)) deb(i, graph[i]);

	auto kek = make_pair(make_pair(12, 34), "abcde");
	vector<int> vec{1,2,3,4};

	deb(kek, vec);

	map<char, int> test;
	test['d'] = 4;
	test['b'] = 2;
	test['a'] = 1;
	test['c'] = 3;

	deb(test, string("xd"));

	vector<bool> boolVec(5);
	boolVec[1] = 1;
	deb(boolVec);

	auto tmp = make_pair(make_pair(0, 1), 2);
	deb(tmp, 3, 4);

	// This doesn't compile, because of order...
	// pair<vi, pii> nested = make_pair(vi{1,2,3}, pii(4,5));
	// deb(nested);

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-value"

	// This doesn't parse properly anymore.
	deb((0,1,2),(3,4,5));

	// This exotic case doesn't work
	struct WeirdStruct { DBP((0,1,2)); };
	WeirdStruct weird;
	deb(weird);

#pragma GCC diagnostic pop

	bitset<16> bitSet;
	bitSet.set(5);
	deb(bitSet);

	vi emptyVec;
	deb(emptyVec);

	cout << string("abcdef") << endl;
	return 0;
}
