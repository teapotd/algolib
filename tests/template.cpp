#include "testing.h"
#include "../lib/template.h"

struct Vert {
	int weight;
	Vi edges;

	DBP(weight, edges);
};

vector<Vert> graph;

int arr[10];

int main() {
	graph.resize(10);

	each(v, graph) {
		v.weight = r(1, 5);
		v.edges.resize(r(0, 9));
		each(x, v.edges) x = r(0, 9);
	}

	rep(i, 0, sz(graph)) deb(i, graph[i]);
	cerr << endl;

	auto kek = mp(mp(12, 34), "abcde");
	vector<int> vec{1,2,3,4};

	deb(kek, vec);

	map<char, int> test;
	test['d'] = 4;
	test['b'] = 2;
	test['a'] = 1;
	test['c'] = 3;

	deb(test, string("xd"));

	vector<bool> boolVec(5);
	deb(boolVec);

	cout << string("abcdef") << endl;

	int i = 5;
	each(x, arr) x = i++;
	const int* constArr = arr;
	deb(span(constArr, 5));

	deb(mp(mp(0, 1), 2), 3, 4);
	return 0;
}
