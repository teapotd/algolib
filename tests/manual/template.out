[: 21, :] graph: [{weight: 100, edges: [0, 1]}, {weight: 101, edges: [1, 2, 3]}, {weight: 102, edges: [2, 3, 4, 5]}, {weight: 103, edges: [3, 4, 5, 6, 7]}, {weight: 104, edges: [4, 5, 6, 7, 8, 9]}, {weight: 105, edges: [5, 6, 7, 8, 9, 10, 11]}, {weight: 106, edges: [6, 7, 8, 9, 10, 11, 12, 13]}, {weight: 107, edges: [7, 8, 9, 10, 11, 12, 13, 14, 15]}, {weight: 108, edges: [8, 9, 10, 11, 12, 13, 14, 15, 16, 17]}, {weight: 109, edges: [9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]}]
[: 22, :] i: 0, graph[i]: {weight: 100, edges: [0, 1]}
[: 22, :] i: 1, graph[i]: {weight: 101, edges: [1, 2, 3]}
[: 22, :] i: 2, graph[i]: {weight: 102, edges: [2, 3, 4, 5]}
[: 22, :] i: 3, graph[i]: {weight: 103, edges: [3, 4, 5, 6, 7]}
[: 22, :] i: 4, graph[i]: {weight: 104, edges: [4, 5, 6, 7, 8, 9]}
[: 22, :] i: 5, graph[i]: {weight: 105, edges: [5, 6, 7, 8, 9, 10, 11]}
[: 22, :] i: 6, graph[i]: {weight: 106, edges: [6, 7, 8, 9, 10, 11, 12, 13]}
[: 22, :] i: 7, graph[i]: {weight: 107, edges: [7, 8, 9, 10, 11, 12, 13, 14, 15]}
[: 22, :] i: 8, graph[i]: {weight: 108, edges: [8, 9, 10, 11, 12, 13, 14, 15, 16, 17]}
[: 22, :] i: 9, graph[i]: {weight: 109, edges: [9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]}
[: 27, :] kek: ((12, 34), abcde), vec: [1, 2, 3, 4]
[: 35, :] test: [(a, 1), (b, 2), (c, 3), (d, 4)], string("xd"): xd
[: 39, :] boolVec: [0, 1, 0, 0, 0]
[: 41, :] mp(mp(0, 1), 2): ((0, 1), 2), 3: 3, 4: 4
[: 46, :] (0,1,2): 2,(3,4,5): 5
[: 51, :] weird: {(0: 2}
[: 57, :] bitSet: 0000000000100000
[: 60, :] emptyVec: []
